modules/linear_boltzmann_solvers/lbs_solver/iterative_methods/ags_solver.cc

```diff
@@ -0,0 +1,121 @@
+// SPDX-FileCopyrightText: 2024 The OpenSn Authors <https://open-sn.github.io/opensn/>
+// SPDX-License-Identifier: MIT
+
+#include "modules//linear_boltzmann_solvers/lbs_solver/iterative_methods/ags_solver.h"
+#include "modules/linear_boltzmann_solvers/lbs_solver/lbs_solver.h"
+#include "framework/math/linear_solver/linear_matrix_action_Ax.h"
+#include "framework/mesh/mesh_continuum/mesh_continuum.h"
+#include "framework/math/petsc_utils/petsc_utils.h"
+#include "framework/logging/log.h"
+#include "framework/runtime.h"
+#include "caliper/cali.h"
+#include <petscksp.h>
+#include <iomanip>
+
+namespace opensn
+{
+namespace lbs
+{
+
+void
+AGSSolver::Solve()
+{
+  CALI_CXX_MARK_SCOPE("AGSSolver::Solve");
+
+  auto grid_ptr = GetCurrentMesh();
+  auto& cell_transport_views = lbs_solver_.GetCellTransportViews();
+  auto& groupsets = lbs_solver_.Groupsets();
+  auto& phi_new = lbs_solver_.PhiNewLocal();
+  auto num_moments = lbs_solver_.NumMoments();
+  std::vector<double> phi_old(lbs_solver_.PhiOldLocal().size(), 0.0);
+
+  auto piecewise_change =
+    [&grid_ptr, &cell_transport_views, &groupsets, num_moments, &phi_new, &phi_old]()
+  {
+    double pw_change = 0.0;
+    for (const auto& cell : grid_ptr->local_cells)
+    {
+      auto& transport_view = cell_transport_views[cell.local_id_];
+      for (int i = 0; i < cell.vertex_ids_.size(); ++i)
+      {
+        for (auto groupset : groupsets)
+        {
+          int gsi = groupset.groups_.front().id_;
+          for (int g = 0; g < groupset.groups_.size(); ++g)
+          {
+            size_t m0g_idx = transport_view.MapDOF(i, 0, gsi + g);
+            double max_phi = std::max(fabs(phi_new[m0g_idx]), fabs(phi_old[m0g_idx]));
+            for (int m = 0; m < num_moments; ++m)
+            {
+              size_t mng_idx = transport_view.MapDOF(i, m, gsi + g);
+              double delta_phi = std::fabs(phi_new[mng_idx] - phi_old[mng_idx]);
+              if (max_phi >= std::numeric_limits<double>::min())
+                pw_change = std::max(delta_phi / max_phi, pw_change);
+              else
+                pw_change = std::max(delta_phi, pw_change);
+            } // for g
+          }   // for m
+        }     // for groupset
+      }       // for i
+    }         // for c
+    double global_pw_change = 0.0;
+    mpi_comm.all_reduce<double>(pw_change, global_pw_change, mpi::op::max<double>());
+    return global_pw_change;
+  };
+
+  // Save qmoms to be restored after each iteration. This is necessary for multiple ags iterations
+  // to function and for keigen-value problems
+  const auto saved_qmoms = lbs_solver_.QMomentsLocal();
```

Comment:
This is a holdover from the previous implementation. I haven't actually tried to run without it. I'll have to dig a little deeper to see what's going on here.

Reply:
@wdhawkins I will let you resolve this conversation when you are ready. In the meantim, I will approve the changes

Reply:
I've looked into this, and, yes, it is necessary. The core of the AGS solver is essentially a loop over groupets that invokes the WGS solver for each groupset. In a nutshell, the WGS solver overwrites `q_moments_local`, and it needs to be restored after the loop over WGS solvers and before the next AGS iteration. Since the core of PI is the invocation of the AGS solver, the same logic applies.